api/python/cellxgene_census/src/cellxgene_census/experimental/ml/geneformer_tokenizer.py

```diff
@@ -0,0 +1,175 @@
+
+        if not (token_dictionary_file and gene_median_file):
+            try:
+                import geneformer
+            except ImportError:
+                # pyproject.toml can't express Geneformer git+https dependency
+                raise ImportError(
+                    "Please install Geneformer with: "
+                    "pip install git+https://huggingface.co/ctheodoris/Geneformer@39ab62e"
+                ) from None
+            if not token_dictionary_file:
+                token_dictionary_file = geneformer.tokenizer.TOKEN_DICTIONARY_FILE
+            if not gene_median_file:
+                gene_median_file = geneformer.tokenizer.GENE_MEDIAN_FILE
+        with open(token_dictionary_file, "rb") as f:
+            gene_token_dict = pickle.load(f)
+        with open(gene_median_file, "rb") as f:
+            gene_median_dict = pickle.load(f)
+
+        # compute model_gene_{ids,tokens,medians} by joining genes_df with Geneformer's
+        # dicts
+        model_gene_ids = []
+        model_gene_tokens = []
+        model_gene_medians = []
+        for gene_id, row in genes_df.iterrows():
+            ensg = row["feature_id"]  # ENSG... gene id, which keys Geneformer's dicts
+            if ensg in gene_token_dict:
+                model_gene_ids.append(gene_id)
+                model_gene_tokens.append(gene_token_dict[ensg])
+                model_gene_medians.append(gene_median_dict[ensg])
+        self.model_gene_ids = np.array(model_gene_ids, dtype=np.int64)
+        self.model_gene_tokens = np.array(model_gene_tokens, dtype=np.int64)
+        self.model_gene_medians = np.array(model_gene_medians, dtype=np.float64)
+
+        assert len(np.unique(self.model_gene_ids)) == len(self.model_gene_ids)
+        assert len(np.unique(self.model_gene_tokens)) == len(self.model_gene_tokens)
+        assert np.all(self.model_gene_medians > 0)
+        # Geneformer models protein-coding and miRNA genes, so the intersection should
+        # be somewhere a little north of 20K.
+        assert len(self.model_gene_ids) > 20_000
+
+    def __enter__(self) -> "GeneformerTokenizer":
+        super().__enter__()
+        # On context entry, load the necessary cell metadata (obs_df)
+        obs_column_names = list(self.obs_attributes)
+        if "soma_joinid" not in self.obs_attributes:
+            obs_column_names.append("soma_joinid")
+        self.obs_df = self.obs(column_names=obs_column_names).concat().to_pandas().set_index("soma_joinid")
+        return self
+
+    def cell_item(self, cell_joinid: int, cell_Xrow: scipy.sparse.csr_matrix) -> Dict[str, Any]:
+        """
+        Given the expression vector for one cell, compute the Dataset item providing
+        the Geneformer inputs (token sequence and metadata).
+        """
+        # project cell_Xrow onto model_gene_ids
+        model_counts = cell_Xrow[:, self.model_gene_ids]
```

Comment:
Hard to tell the impact on performance, but this may not be taking advantage of sparsity, unless scipy  during the multiplication step below is only using the non-zero values, which I doubt.

In other words we may be doing arithmetic operations over the full 20K genes when in reality most of those are 0s which can be left out during multiplication. For handful of cells that may not be a huge deal, but maybe it is impactful over millions of cells.

For reference see my use of csr's `data` slot for this operation in my prototype

https://gist.github.com/pablo-gar/7cbba76caa3f3cfb7b5ebf96b37d2644#file-census_adata_tokenizer-py-L43

Reply:
I think this is OK -- I added some assertions to confirm `model_counts` and `model_expr` remain sparse i.e. their `data` vectors contain no explicit zeroes. (Assertions checked out but are now commented out)

Then, [here is the scipy.sparse code](https://github.com/scipy/scipy/blob/ece199257f857f8a2418587433cfb8b4030a19e1/scipy/sparse/_compressed.py#L467-L469) for efficient multiplication of a `cs{r,c}_matrix` times a dense vector.